Vallox MV Series

bjoernbrings opened the pull request:
Signed-off-by: Björn Brings <bjoernbrings@web.de> (github: bjoernbrings)

This binding is designed to connect to the web interface of Vallox MV series of ventilation unit.
It has been tested so far only with Vallox 350 MV.

In contrast to PR #1560 it does not handle modbus but the newer series with online interface. As several parameters have changed I hand this PR in seperately.

martinvw commented:
@bjoernbrings thanks for explaining the difference and of course for your contribution!

martinvw commented:
Please include the binding in the parent `pom.xml` in that case it will already be tested and analysed by travis. And it will become available in [jfrog builds list](https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/binding/).

Please also check:

http://docs.openhab.org/developers/development/bindings.html
http://docs.openhab.org/developers/development/guidelines.html

And it we would be great if you could also review the report of [the Static code analysis](https://docs.openhab.org/developers/development/bindings.html#static-code-analysis), this also saves me some time while reviewing.

bjoernbrings commented:
Updated pom.xml. Static code analysis did run quite smooth.
But I fail in analysing the Jenkins problem...

martinvw commented:
> But I fail in analysing the Jenkins problem...

No problem its not caused by you.

triller-telekom commented:
I just looked through the list of pull requests and found #1559. Is this PR a duplicate of that?

htreu commented:
Hi @bjoernbrings, thanks for this contribution.
I have some questions regarding your binding and the other Vallox binding from @SuperOok which is in the pipeline (#1560): As I read from this request https://github.com/SuperOok/openhab-vallox/issues/5 the communication of both bindings seem to be not too different. The 1560 version is running RS485 over TCP/IP and you connect over TCP/IP, too. Also the channel-types you defined are very similar to the ones in 1560.
From an enduser point of view it would be nice to have only one binding which handles their devices from a single manufacturer, especially when the device type is the same.
May be you and @SuperOok are able to combine forces and merge your work into a single binding? As #1560 is in the final round of review and nearly done my suggestion would be to wait for it to land in the official repo and then add your support for another thing-type and the websocket communication on top of it. This would even bring @mstormi closer to a solution for his request.

what do you guys think?

bjoernbrings commented:
Happy to discuss how we could merge those two. But those bindings are for different series (SuperOok's is for the old one and mine is for the new MV one). Its not only the type of connection. So parameters are handled differently (e.g. ventilation speed is in percent 0-100 for MV and in steps 0-8 for the old one).
Nevertheless it should be possible to merge those two, but I didn't feel (after reviewing SuperOok's PR) that I could do it as it has quite some complexity and different channels.

Anyways I'm currently struggeling a bit with (my implementation) of the socket connection with jetty as it results in a memory leak (compare community discussion [https://community.openhab.org/t/new-binding-vallox-mv-ventilation-unit-series-beta/36735/12](url)). If anyone has suggestions towards this issue I'd be really happy.

mstormi commented:
From a user perspective, you need both of these two bindings: either you've got an older or low-end box, then you'll need to somehow build(!) and attach RS485<->TCP/IP interface hardware first and use the #1560 binding. If you've got a new (MV series) box, that HW is essentially included but #1560 does not work with it, so you need this new one. While a merge would make sense long-term, I don't see this happen anytime soon, so I clearly vote for a short term oriented approval.
Maybe we should _rename_ this new one to valloxmv or some such, and include a clear, **common** text block in the binding docs to state a) which HW it applies to and b) include the reference to the other binding.
That will prevent foreseeable confusion in userland.

majherek commented:
As a user of Vallox 96SE I use the old one binding from #1560. The serial converter to RS485 to TCP/IP is very simple. You need on use USB RS485 (about $1.5 from eg. aliexpress) and use set2net package with configuration like this: 

`1026:raw:0:/dev/ttyUSB0:9600 8DATABITS NONE 1STOPBIT`.

For me the best way is to rename this binding, like suggest @mstormi.

htreu commented:
@mstormi I'm not into technical details of both device generations nor am I a user of Vallox devices. But from an enduser point of view when searching for a binding for my Vallox box I'd be confused to find more then one and then have to figure out which one is for me. Installing the one and only binding would make sense to me, then choosing which device and connection type I need.
As @bjoernbrings is open for discussion I would just suggest to wait for #1560 to be approved and merged and then have a detailed look on how & whether this one can be integrated.
Besides, I like the proposal for a name change here, since valloxonline suggests some sort of remote/cloud connection.

bjoernbrings commented:
Like proposed I updated the binding name to valloxmv.
Hopefully #1560 is not stale ....

mstormi commented:
Given that the binding was renamed and well-tested, would you mind proceeding so we get this to be an official binding for 2.3 ?

bjoernbrings commented:
Hey Henning, thanks for your feedback. I will try to take care of it over the weekend.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,151 @@
+        <item-type>Number</item-type>
+        <label>Cell state</label>
+        <description>Current cell state (0=heat recovery, 1=cool recovery, 2=bypass, 3=defrosting)</description>
+        <state min="0" max="3" step="1" readOnly="true"></state>
```
Pull down is not shown in PaperUI (just number)
```
	<channel-type id="cellstate">
		<item-type>Number</item-type>
		<label>Cell state</label>
		<description>Current cell state</description>
		<state readOnly="true">
			<options>
				<option value="0">Heat recovery</option>
				<option value="1">Cool recovery</option>
				<option value="2">Bypass</option>
				<option value="3">Defrosting</option>
			</options>
		</state>
	</channel-type>
```

htreu replied:
thats strange, it works for me just like the `state` channel.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,37 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.8
+Bundle-SymbolicName: org.openhab.binding.valloxmv;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.3.0.qualifier
```
Ok

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/README.md:
```diff
@@ -0,0 +1,684 @@
+
+There are even more values supported by the interface but the extraction for most of them has not been implemented so far.
+    
+* A\_CYC\_APPL\_SW\_VERSION   0
```
Ok

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,37 @@
+Bundle-SymbolicName: org.openhab.binding.valloxmv;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.3.0.qualifier
+Export-Package: 
```
Ok

htreu replied:
Yes, this implies that all classes move to the `org.openhab.binding.valloxmv.internal` package. Also the `org.openhab.binding.valloxmv*` entries from the `Import-Package` section must be removed. In addition when I tested it on your code I had to add an extra `org.osgi.framework` entry to the `Import-Package` section.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/ValloxMVBindingConstants.java:
```diff
@@ -0,0 +1,1323 @@
+    // Map MAPADRESSMODBUS maps Vallox names to Modbus addresses, mapAdressReturnBytearray maps Vallox names to byte
+    // position in data return statement
+
+    public static final Map<String, Integer> MAPADRESSMODBUS;
```
Ok -> Exported this to dev readme as well

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVHandler.java:
```diff
@@ -0,0 +1,141 @@
+     * Wait time for the creation of Item-Channel links in seconds. This delay is needed, because the Item-Channel
+     * links have to be created before the thing state is updated, otherwise item state will not be updated.
+     */
+    private static final int WAIT_TIME_CHANNEL_ITEM_LINK_INIT = 1;
```
Hmm, isn't refresh linked to concrete channels?
The following code is handling all channel updates. Any binding where this is done like you suggest (to learn by example)?

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVHandler.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Override
+    public void updateState(ChannelUID uid, State dt) {
```
I did overwrite it to make it public an access it from other objects. How should I handle this?

htreu replied:
Yes, you are right. Didn't see the call from the websocket listener. You may reduce visibility to `protected` though.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+    public ValloxMVWebSocket(ValloxMVHandler voHandler, String ip) {
+        this.voHandler = voHandler;
+        this.ip = ip;
+        logger = LoggerFactory.getLogger(ValloxMVWebSocket.class);
```
Ok

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+        private ChannelUID channelUID;
+
+        public ValloxMVWebSocketListener(ChannelUID channelUID, String updateState) {
+            super();
```
Ups

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+        @OnWebSocketError
+        public void onError(Throwable cause) {
+            voHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            logger.error("Error: {}", cause.getMessage().toString());
```
Ok

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+                // logger.debug("Cellstate: {}, Uptime {}Y, {}h, Current Updtime {}h", bdCellstate, bdUptimeYears,
+                // bdUptimeHours, bdUptimeHoursCurrent);
+
+                // for (String key : ValloxMVBindingConstants.MAPADRESSRETURNBYTEARRAY.keySet()) {
```
Moved to Dev Readme

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+                updateChannel(ValloxMVBindingConstants.CHANNEL_UPTIME_HOURS_CURRENT,
+                        new DecimalType(bdUptimeHoursCurrent));
+
+                // logger.debug("Humidity: {}", bdHumidity);
```
Ok

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+                // getNumber(bytes, ValloxMVBindingConstants.MAPADRESSRETURNBYTEARRAY.get(key)));
+                // }
+            } catch (Exception e) {
+                logger.error("Error: {}", e.getStackTrace().toString());
```
Ok

bjoernbrings commented on addons/binding/pom.xml:
```diff
@@ -118,7 +118,8 @@
     <module>org.openhab.binding.urtsi</module>
     <module>org.openhab.binding.velbus</module>
-    <module>org.openhab.binding.vitotronic</module>
+	<module>org.openhab.binding.valloxmv</module>
```
Ok

bjoernbrings commented:
Hi Henning,
managed to go though your comments already by today. Most of the make sense to me and are easy to change (commented with "ok") and already adapted in my dev environment.
In Things.xml I didn't get the list working (state->translation, cellstate->list) and I have some questions regarding the implementation of the initialization of the request service (react on refresh as you suggested).
If you could advise me I'd be happy to implement like you suggested.
Regards Björn

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVHandler.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Override
+    public void updateStatus(ThingStatus ts, ThingStatusDetail statusDetail) {
```
same, reduce visibility to `protected`.

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVHandler.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Override
+    public void updateStatus(ThingStatus ts) {
```
same, reduce visibility to `protected`.

htreu commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,151 @@
+	<channel-type id="state">
+		<item-type>Number</item-type>
+		<label>State</label>
+		<description>Current state of ventilation unit (1=Fireplace, 2=Away, 3=At home, 4=Boost)</description>
```
I'm figuring out how translation is supposed to work, stay tuned.

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+    public void request(ChannelUID channelUID, String updateState) {
+        try {
+            socket = new ValloxMVWebSocketListener(channelUID, updateState);
+            client.start();
```
I get an NPE here when no Vallox unit is present (just configured one in Paper UI). The `onError` method gets called (due to unsuccessful connection attempt) which in turn calls `close` which destroys the client. On the next iteration NPE strikes.

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+        } catch (IOException e) {
+            logger.error("Error: {}", e.getMessage().toString());
+        } catch (Exception e) {
+            logger.error("Error: {}", e.getMessage().toString());
```
I got NPEs here. Please use something like this: `logger.warn("Error connecting vallox unit.", e);`

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+            client.connect(socket, destUri, request);
+            socket.awaitClose(10, TimeUnit.SECONDS);
+        } catch (IOException e) {
+            logger.error("Error: {}", e.getMessage().toString());
```
same here.

htreu commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/handler/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,357 @@
+        try {
+            client.stop();
+        } catch (Exception e) {
+            logger.error("Error: {}", e.getMessage().toString());
```
same here. You may also just put `e.etMessage()` (which was `null` in my case)  and have a log output of `Error: null` if no message is provided.

bjoernbrings commented:
Implemented the refresh reaction. It's not as easy as for the weather binding as we are writing information back and do not only receive data and have many channels but I hope it is fine like I implemented it.

Any news on how to get the translation working? Thanks.

bjoernbrings commented:
@htreu May I ask you how to proceed? Thanks :-)

htreu commented:
Sorry for the delay. I did configure my firefox to have "de" as the default language. On my machine it works when the `valloxmv_de_DE.properties` is renamed to `valloxmv_de.properties`.

Otherwise the binding looks good from my side.

htreu commented:
@martinvw its your turn, have fun :-)

htreu commented:
@bjoernbrings what I just spotted: there is one commit in your history which is not signed off. You may squash everything (or just repair this single commit) and force push again.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,34 @@
+Bundle-Version: 2.4.0.qualifier
+Import-Package: 
+ javax.measure,
+ javax.measure.quantity;version="1.0.0",
```
Is this version number, needed ideally it should be left out.

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/README.md:
```diff
@@ -0,0 +1,62 @@
+# valloxmv Binding
```
Can it have a more sexy name at this title

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVBindingConstants.java:
```diff
@@ -0,0 +1,137 @@
+    private static final String BINDING_ID = "valloxmv";
+
+    // List of all Thing Type UIDs
+    // @SuppressWarnings("null")
```
Remove this if its not longer needed

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+public class ValloxMVHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(ValloxMVHandler.class);
+    ScheduledFuture<?> updateTasks;
```
Can / should these be private?

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        super(thing);
+    }
+
+    @SuppressWarnings("null")
```
What are we suppressing here, is there really no other way or is it a false positive

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+
+    private final Logger logger = LoggerFactory.getLogger(ValloxMVHandler.class);
+    ScheduledFuture<?> updateTasks;
+    ValloxMVWebSocket vows_send;
```
Please follow the naming guidelines of Java so this would be `vowsSend` or `valloxSendSocket` because I think vows is not really self explanatory

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+                            || (cmd == ValloxMVBindingConstants.STATE_ATHOME)
+                            || (cmd == ValloxMVBindingConstants.STATE_AWAY)
+                            || (cmd == ValloxMVBindingConstants.STATE_BOOST)) {
+                        logger.debug("Changing state to: {}", command.toString());
```
You don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

Please check all

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        } else {
+            if (channelUID.getId().equals(ValloxMVBindingConstants.CHANNEL_STATE)) {
+                try {
+                    int cmd = Integer.parseInt(command.toString());
```
Given it can only be one of those 4 values I think the logic to try to parse it and if it fails ignore is not great...

bjoernbrings replied:
Wanted to make sure, that only expected values are handled further (refresh kicked me out in development which I handle now right before)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+                vows_send.request(null, null);
+            }
+        } else {
+            if (channelUID.getId().equals(ValloxMVBindingConstants.CHANNEL_STATE)) {
```
Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+                    return;
+                }
+            } else if (channelUID.getId().equals(ValloxMVBindingConstants.CHANNEL_ONOFF)) {
+                if (command.toString() == "ON") {
```
This won't work in Java, change it to `"ON".equals(command.toString())` or better compare with the actual command that is an enum so supports comparing with `==` please check other bindings for examples

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        String ip = getConfigAs(ValloxMVConfig.class).getIp();
+        vows_send = new ValloxMVWebSocket(ValloxMVHandler.this, ip);
+
+        BigDecimal bdUpdateInterval = getConfigAs(ValloxMVConfig.class).getUpdateinterval();
```
Maybe another type then a `BigDecimal` for your updateInterval (note the capital I) would make code here way easier and more reabable.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+
+        try {
+            scheduleUpdates();
+        } catch (Exception e) {
```
Please don't do this, why would this be needed here?

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        String ip = getConfigAs(ValloxMVConfig.class).getIp();
+        logger.debug("Connecting to ip: {}", ip);
+        // Open WebSocket
+        ValloxMVWebSocket vows = new ValloxMVWebSocket(ValloxMVHandler.this, ip);
```
maybe name `vows` `webSocket` it will help making code clearer

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        // Open WebSocket
+        ValloxMVWebSocket vows = new ValloxMVWebSocket(ValloxMVHandler.this, ip);
+
+        updateTasks = scheduler.scheduleWithFixedDelay(new Runnable() {
```
Please update this to the Java 8 way of doing this:

```java
updateTasks = scheduler.scheduleWithFixedDelay(() -> {
    // Do a pure read request to websocket interface
    vows.request(null, null);
}, 0, updateInterval, TimeUnit.SECONDS);

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,179 @@
+        super.updateStatus(ts);
+
+    }
+}
```
Please add newlines to the end of all files, some (git) tools have problems if files do not contains a newline after the last data.

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+ * @author Björn Brings - Initial contribution
+ */
+public class ValloxMVWebSocket {
+    private String ip;
```
Can / should be `final` (that is at least a best practice)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+ */
+public class ValloxMVWebSocket {
+    private String ip;
+    private ValloxMVHandler voHandler;
```
Can / should be `final` (that is at least a best practice)

martinvw commented on addons/binding/pom.xml:
```diff
@@ -120,12 +120,13 @@
     <module>org.openhab.binding.tplinksmarthome.test</module>
     <module>org.openhab.binding.urtsi</module>
     <module>org.openhab.binding.velbus</module>
+    <module>org.openhab.binding.valloxmv</module>
```
Please put it on place up.

martinvw replied:
To have the binding being picked up by the distro, you furthermore need to add it to the feature.xml, again at the alphabetically correct position. If you have a dependency on some transport bundle (e.g. upnp, mdns or serial), make sure to add a line for this dependency as well (see the other bindings as an example).

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+            socket.awaitClose(2, TimeUnit.SECONDS);
+        } catch (IOException e) {
+            connectionError(e);
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw replied:
If you catch all `Exception` you are potentially also catching for example `NullPointerException`s and  they tend to miss a message, so you will be printing the message `null` which will make it troublesome to debug, in this case it is always better to log the complete stacktrace.

bjoernbrings replied:
org.eclipse.jetty.websocket.client.WebSocketClient start/stop throw Exception which I have to handle ...

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+    }
+
+    public void connectionError(IOException e) {
+        logger.error("Error connecting vallox unit.", e);
```
If this is caused by an `IOException` and you also update the state, debug is really enough for this, see also:

In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+    public class ValloxMVWebSocketListener {
+        private final CountDownLatch closeLatch = new CountDownLatch(1);
+
+        private Logger logger = LoggerFactory.getLogger(ValloxMVWebSocket.class);
```
Please provide the correct class, `ValloxMVWebSocketListener` afaik

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+    }
+
+    @WebSocket
+    public class ValloxMVWebSocketListener {
```
Should it be an instance class, or would a static inner class be good enough?

bjoernbrings replied:
I'm accessing the ValloxMVHandler from the outter class, that's why I went for the inner class.
(With static inner classs you mean an static outer class, I assume)

martinvw replied:
Okay then maybe you do need it :-)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+                session.getRemote().sendBytes(buf);
+            } catch (IOException e) {
+                connectionError(e);
+            } catch (Exception e) {
```
Please leave this out

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+         * @param hmParameters HashMap for setting data with register as key and value as value
+         * @return ByteBuffer to be sent to websocket
+         */
+        public ByteBuffer generateCustomRequest(Integer mode, HashMap<Integer, Integer> hmParameters) {
```
For method parameters, return types, fields etc, always use the interface of the Java collection types instead of specific implementation, in this case just use `Map<...>`

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+                // requestData (Length 3, Command to get data 246, empty set, checksum [sum of everything before])
+                return generateCustomRequest(246, new HashMap<Integer, Integer>());
+            }
+            if (channelUID.getId().equals(ValloxMVBindingConstants.CHANNEL_STATE)) {
```
Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+                } else if (Integer.parseInt(updateState) == ValloxMVBindingConstants.STATE_BOOST) {
+                    // 30 Min boost (Length 6, Command to set data 249, CYC_BOOST_TIMER (4612) = 30, CYC_FIREPLACE_TIMER
+                    // (4613) = 0, checksum)
+                    HashMap<Integer, Integer> request = new HashMap<Integer, Integer>();
```
See remark above about Map, please apply everywhere

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+        @OnWebSocketError
+        public void onError(Throwable cause) {
+            voHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            logger.debug("Error: {}", cause.getMessage().toString());
```
Please provide context and note that you don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

bjoernbrings replied:
Changed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+        public void onError(Throwable cause) {
+            voHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            logger.debug("Error: {}", cause.getMessage().toString());
+            ValloxMVWebSocket.this.close();
```
What is the consequence of this, it seems that this will cripple this object and its parent, will it be retried / discarded or whatever?

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+                voHandler.dataUpdated();
+            } catch (IOException e) {
+                connectionError(e);
+            } catch (Exception e) {
```
See above, check others

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+            } catch (Exception e) {
+                logger.error("Error: {}", e.getMessage());
+            }
+            logger.debug("Print final");
```
Maybe make this message more production like

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+        private void updateChannel(String strChannelName, State state) {
+            Channel channel = voHandler.getThing().getChannel(strChannelName);
+            if (channel != null) {
+                voHandler.updateState(channel.getUID(), state);
```
AFAIR some bindings use some static value here, maybe pointing immediately to the UID, it seems way more optimal than mapping strings to channels to uid's every time.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+        }
+
+        private BigDecimal getNumber(byte[] bytes, int pos) {
+            return (new BigDecimal(bytes[pos] & 0xff)).multiply(new BigDecimal(256))
```
Is this really the best way, doubles are for this kind of calculations pretty okay as well

bjoernbrings replied:
You mean data type long, don't you? Could rewrite it if you want so, but I'm using BigDecimal everywhere...

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+
+        @OnWebSocketClose
+        public void onClose(int statusCode, String reason) {
+            logger.debug("WebSocket Closed. Code: {}", statusCode);
```
Maybe its wist to also log the `reason`

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,386 @@
+            // Fetch 2 byte number out of bytearray representing the temperature in milli degree kelvin
+            BigDecimal bdTempMiliKelvin = getNumber(bytes, pos);
+            // Return number converted to degree celsius
+            return bdTempMiliKelvin.divide(new BigDecimal(100))
```
Can the unit framework maybe help us in doing this?

htreu replied:
here we go:

```
new QuantityType<>(bdTempMiliKelvin, MetricPrefix.MILLI(SmartHomeUnits.KELVIN)).toUnit(SIUnits.CELSIUS);
```

bjoernbrings commented:
Hey @martinvw, thanks for your feedback. I will take a deeper look next week as I'm traveling over the weekend. After a first look at your comments most of them are pretty clear and straight forward to implement.

bjoernbrings commented:
@martinvw Are you fine with my changes?

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,172 @@
+        String ip = getConfigAs(ValloxMVConfig.class).getIp();
+        valloxSendSocket = new ValloxMVWebSocket(ValloxMVHandler.this, ip);
+
+        BigDecimal bdUpdateInterval = getConfigAs(ValloxMVConfig.class).getUpdateinterval();
```
IMHO BigDecimal's are not the proper type for such things / calculations. You don't need decimals and you don't need big.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandlerFactory.java:
```diff
@@ -0,0 +1,53 @@
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(THING_TYPE_SAMPLE)) {
```
Can you rename `THING_TYPE_SAMPLE`?

Position literals first in `String` comparisons - that way if the `String` is `null` you won't get a `NullPointerException`, it'll just return `false`. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+        try {
+            client.stop();
+        } catch (IOException e) {
+            connectionError(e);
```
A connection error when closing can be ignore or maybe logged on debug, more is not needed

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+        } catch (IOException e) {
+            connectionError(e);
+        } catch (Exception e) {
+            logger.error("Error: {}", e);
```
Always provide a proper context, but imho nothing more then debug logging is needed.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+        } catch (IOException e) {
+            connectionError(e);
+        } catch (URISyntaxException | InterruptedException e) {
+            logger.error("Error: {}", e);
```
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please change this to use the correct logging levels and if possible change the ONLINE/OFFLINE status of thing / bridge.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+            connectionError(e);
+        } catch (URISyntaxException | InterruptedException e) {
+            logger.error("Error: {}", e);
+        } catch (Exception e) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+    public class ValloxMVWebSocketListener {
+        private final CountDownLatch closeLatch = new CountDownLatch(1);
+
+        private Logger logger = LoggerFactory.getLogger(ValloxMVWebSocketListener.class);
```
Please make all your logger's final

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+        private final CountDownLatch closeLatch = new CountDownLatch(1);
+
+        private Logger logger = LoggerFactory.getLogger(ValloxMVWebSocketListener.class);
+        private String updateState;
```
It seems that these fields could also be final, that would have my preference.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+                } else if (Integer.parseInt(updateState) == ValloxMVBindingConstants.STATE_AWAY) {
+                    // Away (Length 8, Command to set data 249, CYC_STATE (4609) = 1, CYC_BOOST_TIMER (4612) = 0,
+                    // CYC_FIREPLACE_TIMER (4613) = 0, checksum)
+                    Map<Integer, Integer> request = new HashMap<Integer, Integer>();
```
Would one flow to generate the HashMap and return be enough?

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+                    return generateCustomRequest(249, request);
+                }
+            } else if (channelUID.getId().equals(ValloxMVBindingConstants.CHANNEL_ONOFF)) {
+                HashMap<Integer, Integer> request = new HashMap<Integer, Integer>();
```
If this needs to stay, please change the vars to a Map.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+                    // 15 Min fireplace (Length 6, Command to set data 249, CYC_BOOST_TIMER (4612) = 0,
+                    // CYC_FIREPLACE_TIMER
+                    // (4613) = 15, checksum)
+                    Map<Integer, Integer> request = new HashMap<Integer, Integer>();
```
Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`

Apply everywhere.

martinvw commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVWebSocket.java:
```diff
@@ -0,0 +1,387 @@
+        public void onError(Throwable cause) {
+            voHandler.updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            logger.debug("Connection failed: {}", cause.getMessage());
+            // ValloxMVWebSocket.this.close();
```
Remove this commented out statement.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    protected void updateState(String strChannelName, State dt) {
```
I want to access them from class ValloxMVWebSocket and as they are protected in super class.

kaikreuzer replied:
But they are protected here as well, so what makes the difference...?

bjoernbrings replied:
protected in super class => private in implementation though not accessable from ValloxMVWebSocket
protected in implementation => accessable in ValloxMVWebSocket

kaikreuzer replied:
Ah, thanks, now I understand. The difference is actually that protected makes it available to classes in the same package as well - while it is also protected before, it is only accessible to classes that are in the same package as the BaseThingHandler. So all fine, thanks!

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,170 @@
+
+    @Override
+    public void dispose() {
+        updateTasks.cancel(true);
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        scheduleUpdates();
+    }
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,170 @@
+        if (command instanceof QuantityType) {
+            return (QuantityType<U>) command;
+        }
+        return new QuantityType<U>(new BigDecimal(command.toString()), defaultUnit);
```
"The constructor QuantityType<U>(DecimalType, Unit<U>) is undefined"

kaikreuzer replied:
Then maybe `new QuantityType<U>((Number) command, defaultUnit)` as DecimalType is also a Number?

bjoernbrings replied:
perfect -> corrected
=> we should change it in addons\binding\org.openhab.binding.nest\src\main\java\org\openhab\binding\nest\handler\NestBaseHandler.java and addons\binding\org.openhab.binding.nest\src\main\java\org\openhab\binding\nest\handler\NestThermostatHandler.java as well (from where I have reused the code)

kaikreuzer replied:
Feel free to create a PR with these fixes!

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/README.md:
```diff
@@ -0,0 +1,62 @@
+## Thing Configuration
+
+The Thing needs the information at which IP the web interface could be reached and how often the values should be updated.
+Minimum update interval is limited to 15 sec in order to avoid polling again before results have been evaluated.
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Supported Things
+
+There is one thing supporting the connection via the web interface. There is NO support of former modbus connected devices.
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,198 @@
+		<label>Vallox ventilation unit</label>
+		<description>Interface to online interface of Vallox ventilation unit</description>
+		<channels>
+			<channel id="onoff" typeId="onoff" />
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,198 @@
+			<parameter name="updateinterval" type="integer" required="false">
+				<label>Update interval</label>
+				<description>Data update interval in seconds (default: 60s)</description>
+				<required>true</required>
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,198 @@
+		</channels>
+		<config-description>
+			<parameter name="ip" type="text" required="true">
+				<label>IP</label>
```
corrected and changed to IP Address

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,198 @@
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+	<thing-type id="valloxmv">
+		<label>Vallox ventilation unit</label>
```
corrected

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,11 @@
+	xmlns:binding="http://eclipse.org/smarthome/schemas/binding/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/binding/v1.0.0 http://eclipse.org/smarthome/schemas/binding-1.0.0.xsd">
+
+	<name>Vallox ventilation unit</name>
```
corrected

bjoernbrings commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -516,6 +516,11 @@
         <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.urtsi/${project.version}</bundle>
     </feature>
 
+    <feature name="openhab-binding-valoxmv" description="ValoxMV Binding" version="${project.version}">
```
corrected

bjoernbrings commented:
@kaikreuzer: Thanks for the very quick feedback. Tried to adapt all of it as quick as you commented. I could not solve one comment.

Could you help me with the following problems:
-     "Current cell state" is not shown as pull down (though it is coded exactly the same way as "Current state of ventilation unit" which is shown as pull down)
-     Writeable channels (e.g. Extract fan base speed) switch to -NaN when changed via PaperUI

kaikreuzer commented:
> "Current cell state" is not shown as pull down (though it is coded exactly the same way as "Current state of ventilation unit" which is shown as pull down)

Simple: The difference is that "Current cell state" is marked as readOnly=true, which means that the User cannot set a value and thus a drop down menu would not make sense here.

> Writeable channels (e.g. Extract fan base speed) switch to -NaN when changed via PaperUI

You should change the lines
```
		<state readOnly="false" min="0" max="1" pattern="%d %unit%" />
```
to
```
		<state readOnly="false" min="0" max="100" pattern="%d %%" />
```
for all your percent value channels.

bjoernbrings commented:
You're right, but I'd have expected to see the text and not just a number (that's why I missed the pulldown).

Changed things file according to comment. But I still find it strange as values are displayed correctly but could not be changed.
It seems like values can be changed only once (before page refresh).

kaikreuzer commented:
> It seems like values can be changed only once (before page refresh).

Yes, I noticed the same; seems to be a bug in the Paper UI. When refreshing the page, it is again possible to update the value.

bjoernbrings commented:
Thanks for you review. That was quite quick at the end 👍 

Any idea (or link to a discussion) when a new stable (or monthly) release will come?

kaikreuzer commented:
Next release is probably in autumn or towards the end of the year.
But I am actually planning a milestone release the next days, so this binding will be included in that.
Watch out for news in https://community.openhab.org/c/announcements.

bjoernbrings commented on addons/binding/org.openhab.binding.valloxmv/src/main/java/org/openhab/binding/valloxmv/internal/ValloxMVHandler.java:
```diff
@@ -0,0 +1,192 @@
+                if (quantity == null) {
+                    return;
+                }
+                int milliKelvin = quantity.multiply(new BigDecimal(100)).intValue();
```
It should be called centi-Kelvin. And you're right, for sending updates I already used the UoM code but didn't change it here.
